frappe/permissions.py

```diff
@@ -361,14 +361,37 @@ def has_user_permission(doc, user=None, debug=False, *, ptype=None):
 		# if allowed_docs is empty it states that there is no applicable permission under the current doctype
 
 		# only check if allowed_docs is not empty
-		if allowed_docs and docname and str(docname) not in allowed_docs:
-			# no user permissions for this doc specified
-			debug and _debug_log(
-				"User doesn't have access to this document because of User Permissions, allowed documents: "
-				+ str(allowed_docs)
-			)
-			push_perm_check_log(_("Not allowed for {0}: {1}").format(_(doctype), docname), debug=debug)
-			return False
+		if allowed_docs:
+			condition = True
+			if doc.meta.is_tree and ptype == "create":
+				if parent := doc.get(doc.nsm_parent_field):
+					from frappe.utils.nestedset import get_ancestors_of
+
+					for d in [parent, *get_ancestors_of(doctype, parent)]:
+						if d in allowed_docs:
+							try:
+								up = frappe.get_doc(
+									"User Permission",
+									{"allow": doctype, "for_value": d, "user": user},
+									fields=["hide_descendants"],
+								)
```

Comment:
potentially O(n) calls for deeply nested parent.. consider moving outside loop

Reply:
or even better, we add `has_descendants` here?
https://github.com/frappe/frappe/blob/7ab8a396eda01c268f342f10785584a6e7e91bde/frappe/core/doctype/user_permission/user_permission.py#L116

then it's accessible in the local `user_permissions` variable